### DISCUSSION TITLE
Hi Rector Team, Is there a way to prevent PhpStormStubsMap fatal error in rector 0.14 phpstan phar dependency?

### DISCUSSION THREAD

**Discussion by alexandreelise** [rectorphp/rector#7401] [Q&A] [answered]
Hi, @TomasVotruba ,

Before asking this question I searched in the actual code. I found that in the latest rector 0.14.0 the phpstan.phar dependency has in it a vendor folder that autoloads the classmap and PhpStormStubsMap at the same time which cause cannot redeclare class fatal error. What might be a fix is not to load PhpStormStubsMap.php. 
The offending line is in bootstrap.php at line 25 in phpstan.phar in rector 0.14.0 dependencies.

```
require_once 'phar://' . __DIR__ . '/phpstan.phar/vendor/jetbrains/phpstorm-stubs/PhpStormStubsMap.php';

```

The problem started since rector 0.13
This issue doesn't exists in rector 0.12

That's why, for now, I'm stuck on 0.12. Can you give me a hint on how to solve the issue?

Have a great day

**Comment by TomasVotruba** [comment, marked as answer]
Hi @alexandreelise ,

thanks for report. Which exact commit broke this? That would help us found out what change made this happen and be able to fix it.

**Reply by alexandreelise** [reply]
Hi @TomasVotruba ,

Since this issue seems to be in a phar file. It won't be easy to dig in and find which commit is responsible for this problem. Github gives up when files are to huge. I will have to fetch all the branches since 0.12 and do a find in files using phpstorm. Is this the way you would do it? I can give it a try. I am not an expert at git yet maybe you would use something like git bissect?

**Reply by TomasVotruba** [reply]
What file is this in?

```php
require_once 'phar://' . __DIR__ . '/phpstan.phar/vendor/jetbrains/phpstorm-stubs/PhpStormStubsMap.php';
```

**Reply by TomasVotruba** [reply]
>  I will have to fetch all the branches since 0.12 

That could be a good start. 

I'd go with 0.12.0, then 0.12.23, then 0.12.12... etc. basically cutting down by half that passes. I think in 5 version checks you'll have it with this technique.

**Reply by alexandreelise** [reply]
divide and conquer. dichotomy.

**Reply by alexandreelise** [reply]
The offending line is in bootstrap.php at line 25 in phpstan.phar in rector 0.14.0 dependencies.

this line is exactly in :  
https://github.com/rectorphp/rector/blob/a6435672ab10f734b9cdb548af688afd60bdced4/vendor/phpstan/phpstan/bootstrap.php#L23

Here it is. It wasn't in the phar file

**Reply by TomasVotruba** [reply]
Thanks :+1: 
Seems this file is there last 15 months: https://github.com/phpstan/phpstan/blame/1.8.x/bootstrap.php

We cannot modify PHPStan. So we'll need a reproducible Github repository, that shows your issue. I'm curious what exact setup you have and what exact error do you get.

**Reply by alexandreelise** [reply]
Is it possible to have this issue if another phpstan is installed in the root vendor folder? I think that grumphp that I use requires phpstan

**Reply by TomasVotruba** [reply]
Not sure. AFAIK composer can install only one PHPStan version at a time. 

It starts to sound like some messy autoload issue hack by another package :)

**Comment by alexandreelise** [comment]
Hi Rector Team you can mark this question as solved. I now understand better how rector works.

**Reply by TomasVotruba** [reply]
Hi, I believe you can choose the answer :)
